pkg/cluster/config/encoding/encoding.go

```diff
@@ -1,146 +0,0 @@
-
-Licensed under the Apache License, Version 2.0 (the "License");
-you may not use this file except in compliance with the License.
-You may obtain a copy of the License at
-
-    http://www.apache.org/licenses/LICENSE-2.0
-
-Unless required by applicable law or agreed to in writing, software
-distributed under the License is distributed on an "AS IS" BASIS,
-WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-See the License for the specific language governing permissions and
-limitations under the License.
-*/
-
-// Package encoding implements apiVersion aware functionality to
-// Marshal / Unmarshal / Load Config
-package encoding
-
-import (
-	"bytes"
-	"fmt"
-	"io/ioutil"
-
-	"sigs.k8s.io/yaml"
-
-	"sigs.k8s.io/kind/pkg/cluster/config"
-)
-
-// Load reads the file at path and attempts to load it as a yaml Config
-// after detecting the apiVersion in the file
-// (or defaulting to the current version if none is specified)
-// If path == "" then the default config for the current version is returned
-func Load(path string) (config.Any, error) {
-	if path == "" {
-		return config.New(), nil
-	}
-	// read in file
-	contents, err := ioutil.ReadFile(path)
-	if err != nil {
-		return nil, err
-	}
-	// read in some config version
-	// TODO(bentheelder): we do not use or respect `kind:` at all
-	// possibly we should require something like `kind: "Config"`
-	cfg, err := Unmarshal(contents)
-	if err != nil {
-		return nil, err
-	}
-	return cfg, err
-}
-
-// LoadCurrent is equivalent to Load followed by cfg.ToCurrent()
-func LoadCurrent(path string) (*config.Config, error) {
-	cfg, err := Load(path)
-	if err != nil {
-		return nil, err
-	}
-	return cfg.ToCurrent(), nil
-}
-
-// used to sniff a config for it's api version
-type configOnlyVersion struct {
-	APIVersion string `json:"apiVersion,omitempty"`
-}
-
-// helper to sniff and validate apiVersion
-func detectVersion(raw []byte) (version string, err error) {
-	c := configOnlyVersion{}
-	if err := yaml.Unmarshal(raw, &c); err != nil {
-		return "", err
-	}
-	switch c.APIVersion {
-	// default to the current api version if unspecified, or explicitly specified
-	case config.SchemeGroupVersion.String(), "":
-		return config.SchemeGroupVersion.String(), nil
-	}
-	return "", fmt.Errorf("invalid version: %v", c.APIVersion)
-}
-
-// Unmarshal is an apiVersion aware yaml.UnmarshalStrict for config.Any
-func Unmarshal(raw []byte) (config.Any, error) {
-	if raw == nil {
-		return nil, fmt.Errorf("nil input")
-	}
-	// sniff and validate version
-	version, err := detectVersion(raw)
-	if err != nil {
-		return nil, err
-	}
-	// load version specific config type
-	var cfg config.Any
-	var rawCfg interface{}
-	switch version {
-	case config.SchemeGroupVersion.String():
-		// this allows strict unmarshaling with the version header
-		layout := &struct {
-			configHeader
-			config.Config
-		}{}
-		cfg = &layout.Config
-		rawCfg = layout
-	}
-	err = yaml.UnmarshalStrict(raw, rawCfg)
```

Comment:
this does unfortunately mean we lose strict unmarshalling (See the config issue for dicussion) but it is currently par for the course in kubernetes tooling.

Reply:
strict unmarshal warnings (or errors?) would be pretty easy to add after this PR is merged as a pre-step before Decode.

Reply:
had a look at how to create a custom decoder too. seems doable in the near future.

Reply:
@BenTheElder, considering above comments it seems to me that this can be acceptable for now right?

Reply:
> this can be acceptable for now right

yes,
it's just that i recommend that we enable strict unmarshal with the first kind release.

since we cannot merge a strict decoded in api-machinery in this release, we should probably go for the way kubeadm is doing it. calling `yaml.UnmarshalStrict` before Decode.

Reply:
yes, sorry, I wanted to call out that we had this functionality change, but it makes sense.

we can sort out strict unmarshal again later, I'm ok with that not being in the first "release" since the first release will be very alpha grade and not stable. I'm going to make this clear with the versioning of the APIs and kind itself. 

we should get strict unmarshal before we make compatibility guarantees between versions so it isn't a breaking change later.